jib-core/src/test/java/com/google/cloud/tools/jib/builder/steps/StepsRunnerTest.java

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.tools.jib.builder.steps;
+
+import com.google.cloud.tools.jib.api.DescriptorDigest;
+import com.google.cloud.tools.jib.builder.ProgressEventDispatcher;
+import com.google.cloud.tools.jib.builder.steps.PullBaseImageStep.ImagesAndRegistryClient;
+import com.google.cloud.tools.jib.configuration.BuildContext;
+import com.google.cloud.tools.jib.image.DigestOnlyLayer;
+import com.google.cloud.tools.jib.image.Image;
+import com.google.common.collect.ImmutableList;
+import com.google.common.util.concurrent.ForwardingExecutorService;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.ListeningExecutorService;
+import java.security.DigestException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.Mockito;
+import org.mockito.junit.MockitoJUnitRunner;
+
+/** Tests for {@link StepsRunner}. */
+@RunWith(MockitoJUnitRunner.class)
+public class StepsRunnerTest {
+
+  // ListeningExecutorService is annotated with @DoNotMock, so define a concrete class.
```

Comment:
The instructions seem to indicate maybe a different approach, but I don't know if you already considered that.

> Use TestingExecutors.sameThreadScheduledExecutor, or wrap a real Executor from java.util.concurrent.Executors with MoreExecutors.listeningDecorator

Reply:
I saw the comment, but I think it cannot apply to what I'm trying to achieve. `TestingExecutors.sameThreadScheduledExecutor` and `MoreExecutors.listeningDecorator` actually executes given tasks, whereas here I want to mock it at a higher level to simulate tasks.

I tried quite a lot of time to find the simplest way to mock `ListeningExecutorService.submit()`, but this code was the shortest I could get so far.